src/server/services/redeemableCode.service.ts

```diff
@@ -36,39 +65,255 @@ export async function consumeRedeemableCode({
+              data: {
+                productId: consumedCode.price.product.id,
+                priceId: consumedCode.price.id,
+                status: 'active',
+                currentPeriodStart: now.toDate(),
+                currentPeriodEnd: now
+                  .add(
+                    consumedCode.unitValue,
+                    consumedCode.price.interval as 'day' | 'month' | 'year'
+                  )
+                  .toDate(),
+              },
+            });
+          } else if (consumedTierOrder === membershipTierOrder) {
+            // If it's the same tier, we just extend the current period:
+            await tx.customerSubscription.update({
+              where: { id: activeUserMembership.id },
+              data: {
+                status: 'active',
+                currentPeriodEnd: dayjs(activeUserMembership.currentPeriodEnd)
+                  .add(
+                    consumedCode.unitValue,
+                    consumedCode.price.interval as 'day' | 'month' | 'year'
+                  )
+                  .toDate(),
+              },
+            });
+          }
+        }
+      }
+
+      if (!activeUserMembership) {
+        // Create a new membership:
+        const now = dayjs();
+        await tx.customerSubscription.create({
+          data: {
+            id: `redeemable-code-${consumedCode.code}`,
+            userId: userId,
+            productId: consumedCode.price.product.id,
+            priceId: consumedCode.price.id,
+            status: 'active',
+            currentPeriodStart: now.toDate(),
+            currentPeriodEnd: now
+              .add(consumedCode.unitValue, consumedCode.price.interval as 'day' | 'month' | 'year')
+              .toDate(),
+            cancelAtPeriodEnd: true, // We assume they want to cancel at the end of the period
+            cancelAt: null, // No cancellation date yet
+            metadata: {},
+            createdAt: now.toDate(),
+          },
+        });
+      }
+
+      const consumedProductMetadata = consumedCode.price.product
+        .metadata as SubscriptionProductMetadata;
+
+      const date = dayjs().format('YYYY-MM');
+
+      await withRetries(async () => {
+        // Grant buzz right away:
+        await grantBuzzPurchase({
+          userId: userId,
+          amount: Number(consumedProductMetadata.monthlyBuzz ?? 5000), // Default to 5000 if not specified
+          description: `Membership Bonus`,
+          transactionType: TransactionType.Purchase,
+          externalTransactionId: `civitai-membership:${date}:${userId}:${consumedCode.price!.product.id
+            }`,
+          details: {
+            type: 'membership-purchase',
+            date: date,
+            productId: consumedCode.price!.product.id,
+          },
+        });
+
+        await invalidateSession(userId);
+        await getMultipliersForUser(userId, true);
+        await deliverMonthlyCosmetics({
+          userIds: [userId],
+        })
+      });
```

Comment:
Should this go outside of the transaction? Feels like it has to happen regardless